eos-grub-upgrade

```diff
@@ -0,0 +1,276 @@
+      ;;
+    /dev/*p[0-9])
+      root_disk=${root_part%p?}
+      ;;
+  esac
+
+  if [[ -z ${root_disk} ]]; then
+    echo "No root disk found for ${root_part}"
+    exit 1
+  fi
+
+  echo "${root_disk}"
+}
+
+# Returns "gpt" or "dos"
+get_partition_table() {
+  local root_disk=$1
+
+  local pt_label=$(blkid -o value -s PTTYPE "${root_disk}")
+  if [[ -z ${pt_label} ]]; then
+    echo "blkid -o value -s PTTYPE '${root_disk}' failed"
+    exit 1
+  fi
+
+  echo "${pt_label}"
+}
+
+get_firmware_type() {
+  if [[ -d /sys/firmware/efi ]]; then
+    echo "uefi"
+  else
+    echo "bios"
+  fi
+}
+
+mount_ESP() {
+  local root_disk=$1
+  local esp_guid="C12A7328-F81F-11D2-BA4B-00A0C93EC93B"
+
+  # Check if ESP does exist and retrieve the partition name
+  local esp_part=$(sfdisk -d "${root_disk}" | grep -i type=${esp_guid} | sed 's/ .*//')
+  if [[ -z ${esp_part} ]]; then
+    echo "ESP not found"
+    exit 1
+  fi
+
+  # On Endless-only (not dual boot) UEFI installation the ESP is always the first
+  # partition on the booting disk
+  local esp_partno=$(get_last_char "${esp_part}")
+  if [[ ${esp_partno} != "1" ]]; then
+    echo "ESP is not the first partition (how?)"
+    exit 1
+  fi
+
+  # Mount the ESP
+  if ! mount "${esp_part}" ${ESP_MOUNTPOINT}; then
+    echo "Failed to mount ESP"
+    exit 1
+  fi
+}
+
+check_for_upgrade() {
+  local old=$1
+  local new=$2
+
+  # Check both are regular files
+  if [[ -f ${old} && -f ${new} ]]; then
+
+    # pesign returns 0 when the file is signed, 1 otherwise (also when the file
+    # is not an UEFI application)
+    pesign -S -i "${new}" &> /dev/null
+    local new_signed=$?
+
+    pesign -S -i "${old}" &> /dev/null
+    local old_signed=$?
+
+    # If we are going to upgrade a signed file we must be sure that also the
+    # new file is signed
+    if [[ ${new_signed} -eq 0 && ${new_signed} -ne ${old_signed} ]]; then
+      return
+    fi
+
+    # Compare the two files, if they are different copy the new file into a
+    # temporary ${old}.new file in the same directory (and partition) of the
+    # file to overwrite
+    if ! cmp --silent "${old}" "${new}"; then
+      cp -a "${new}" "${old}".new
+      echo "${old}"
+    fi
+  fi
+}
+
+# Get the installation type. It could embed a complex logic to determine the
+# installation type.
+get_installation_type() {
+  local firmware_type=$1
+  local partition_table=$2
+
+  if [[ ${firmware_type} == "uefi" && ${partition_table} == "gpt" ]]; then
+    local inst_type
+    for inst_type in "${!GRUB_AMD64[@]}"; do
+      if [[ -f ${GRUB_AMD64[$inst_type]} ]]; then
+        echo "${inst_type}"
+        return
+      fi
+    done
+  fi
+}
+
+do_uefi_upgrade() {
+  local root_disk=$1
+  local old_file
+  local new_file
+
+  # Mount the ESP
+  mount_ESP "${root_disk}"
+
+  # Get the installation type
+  local inst_type=$(get_installation_type "uefi" "gpt")
+  if [[ -z ${inst_type} ]]; then
+    echo "Installation type unknown"
+    exit 1
```

Comment:
I don't think it's appropriate to exit failure from this script on systems with unsupported configurations, because that will mean the systemd unit fails, we'll enter a degraded state, and this causes log messages to be visible on shutdown: https://phabricator.endlessm.com/T17983#419572